packages/node-http-handler/src/node-http-handler.spec.ts

```diff
@@ -38,55 +36,88 @@ describe("NodeHttpHandler", () => {
-          httpAgent: new http.Agent({ keepAlive: false, maxSockets: randomMaxSocket }),
-        }),
-      ],
-    ])("sets httpAgent when input is %s", async (_, option) => {
-      const nodeHttpHandler = new NodeHttpHandler(option);
-      await nodeHttpHandler.handle({ protocol: "http:", headers: {}, method: "GET", hostname: "localhost" } as any);
-      expect(hRequestSpy.mock.calls[0][0]?.agent.keepAlive).toEqual(false);
-      expect(hRequestSpy.mock.calls[0][0]?.agent.maxSockets).toEqual(randomMaxSocket);
+      it.each([
+        ["an option hash", { httpsAgent: new https.Agent({ keepAlive: true, maxSockets: randomMaxSocket }) }],
+        [
+          "a provider",
+          async () => ({
+            httpsAgent: new https.Agent({ keepAlive: true, maxSockets: randomMaxSocket }),
+          }),
+        ],
+      ])("sets httpsAgent when input is %s", async (_, option) => {
+        const nodeHttpHandler = new NodeHttpHandler(option);
+        await nodeHttpHandler.handle({ protocol: "https:" } as any);
+        expect(hsRequestSpy.mock.calls[0][0]?.agent.keepAlive).toEqual(true);
+        expect(hsRequestSpy.mock.calls[0][0]?.agent.maxSockets).toEqual(randomMaxSocket);
+      });
     });
 
-    it.each([
-      ["an option hash", { httpsAgent: new https.Agent({ keepAlive: true, maxSockets: randomMaxSocket }) }],
-      [
-        "a provider",
-        async () => ({
-          httpsAgent: new https.Agent({ keepAlive: true, maxSockets: randomMaxSocket }),
-        }),
-      ],
-    ])("sets httpsAgent when input is %s", async (_, option) => {
-      const nodeHttpHandler = new NodeHttpHandler(option);
-      await nodeHttpHandler.handle({ protocol: "https:" } as any);
-      expect(hsRequestSpy.mock.calls[0][0]?.agent.keepAlive).toEqual(true);
-      expect(hsRequestSpy.mock.calls[0][0]?.agent.maxSockets).toEqual(randomMaxSocket);
+    describe("#handle", () => {
+      it("should only generate a single config when the config provider is async and it is not ready yet", async () => {
+        let providerInvokedCount = 0;
+        let providerResolvedCount = 0;
+        const slowConfigProvider = async () => {
+          providerInvokedCount += 1;
+          await new Promise((r) => setTimeout(r, 15));
+          providerResolvedCount += 1;
+          return {
+            connectionTimeout: 12345,
+            socketTimeout: 12345,
+            httpAgent: null,
+            httpsAgent: null,
+          };
+        };
+
+        const nodeHttpHandler = new NodeHttpHandler(slowConfigProvider);
+        const promises = [];
+        for (let i = 0; i < 100; ++i) {
+          promises.push(nodeHttpHandler.handle({} as unknown as HttpRequest));
+        }
+
+        expect(providerInvokedCount).toBe(1);
+        expect(providerResolvedCount).toBe(0);
+        await new Promise((r) => setTimeout(r, 50));
+        expect(providerInvokedCount).toBe(1);
+        expect(providerResolvedCount).toBe(1);
+        await Promise.all(promises);
```

Comment:
await Promise.all() to call handle in parallel while slowConfigProvide is getting resolved.
```suggestion
        await Promise.all(promises);
        expect(providerInvokedCount).toBe(1);
        expect(providerResolvedCount).toBe(1);
```